### ISSUE TITLE
_draw() can be triggered more than once per frame

### ISSUE THREAD

**Issue by vpellen** [godotengine/godot#73338]
### Godot version

v4.0.rc2.official [d2699dc7a]

### System information

Windows 10

### Issue description

tl;dr: the `_draw` queue is being flushed after `_process` _and_ `_physics_process`, rather than just after `_process`.

Longer version: `queue_redraw` queues `_draw` to be called, however that queue is flushed even after physics frames. While this isn't typically a major issue, there could conceivably be situations where physics frames are occurring more frequently than rendering frames, or for both `_physics_process` and `_process` to call `queue_redraw`, which would lead to redundant calls to `_draw` before rendering actually takes place after `_process` is done. Ideally the queue should be flushed only once, after process, prior to rendering.

### Steps to reproduce

This is easiest to observe by sticking a call to queue_redraw inside a _physics_process, clamping the FPS in the project settings to something like 10, and printing out any calls to _process, _physics_process, and _draw.

### Minimal reproduction project

N/A

**Comment by YuriSizov** [comment]
Please, provide an MRP instead of removing that section. This issue template exists for a reason.

**Comment by vpellen** [comment]
> Please, provide an MRP instead of removing that section. This issue template exists for a reason.

I believed this to be an issue so trivially reproducible that an MRP would be superfluous. However, as per your request, I have uploaded an example project. In addition, I have also included the 10 line script which is the heart of the program, should people not wish to download the full 4mb zip file containing nothing but that script attached to a Node2D within a new project.

I can annotate it if necessary.

**Comment by dalexeev** [comment]
> full 4mb zip file containing nothing

You should delete the `.godot` folder before creating an MRP archive. The shader cache takes up most of it.

**Comment by vpellen** [comment]
> You should delete the `.godot` folder before creating an MRP archive. The shader cache takes up most of it.
Good advice, thank you, done.

**Comment by YuriSizov** [comment]
> I believed this to be an issue so trivially reproducible that an MRP would be superfluous. However, as per your request, I have uploaded an example project.

Appreciated. While an issue may be trivial to reproduce by the reporter's account, having an MRP gives both the contributor and the reporter some point of common reference, which is extremely useful. It also allows triagers, who may not be that familiar with any particular area, to confirm that at least as described and as provided the bug is indeed reproduceable. And it shaves off a few annoying minutes of setting up a new project for someone who wants to look into it, which is also always nice.

**Comment by vpellen** [comment]
If it's any consolation, in spite of being indirectly linked to rendering, I doubt this is platform specific - a quick dig through the source indicates about what I expected, which is that CanvasItem just pushes a callback onto the MessageQueue, which I presume gets flushed every inter-frame interval when all the deferred calls get processed.

I don't know if there's a reliable way to postpone a callback until just before rendering. If there is, this may well be a fairly trivial fix. If not, I suspect it's a fundamental architectural issue that won't get solved for a while.
